R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
+#' @export
+get_ci_pois <- function(interval, x, y = 1) {
+  # TODO integrate Poisson CI into function that calculates SMR
+
+  # check validity of inputs
+
+  if (!is.numeric(interval)) {
+    stop("`interval` must be numeric")
+  }
+
+  if (length(interval) != 1) {
+    stop("`interval` must be a scalar")
+  }
+
+  if (interval < 0 | 1 < interval) {
+    stop("`interval` must be between 0 and 1")
+  }
+
+  if (!is.numeric(x)) {
+    stop("`x` must be numeric")
+  }
+
+  if (!all(x %% 1 == 0, na.rm = TRUE)) {
+    stop("`x` must be a vector of integers")
   }
 
-  # determine limit quantiles
+  if (!all(x > 0, na.rm = TRUE)) {
+    stop("`x` must be a vector of positive values")
+  }
+
+  if (!is.numeric(y)) {
+    stop("`y` must be numeric")
+  }
 
+  if (!all(y > 0, na.rm = TRUE)) {
+    stop("`y` must be a vector of positive values")
+  }
+
+  if (length(x) != length(y)) {
+    if (length(y) == 1) {
+      if (y != 1) {
+        stop("expected `y = 1` given length of `x` and `y`") # avoid recycling values other than 1
+      }
+    } else {
+      stop("length of `y` is not compatible with that of `x`")
+    }
+  }
+
+  # determine quantiles of interest
   alpha <- 1 - interval
-  q_low <- alpha / 2
-  q_high <- 1 - (alpha / 2)
-
-  # construct confidence intervals
-
-  if (dist == "normal") {
-    sd <- sqrt(variance)
-    ci <- purrr::map2(estimate, sd, ~ stats::qnorm(c(q_low, q_high), .x, .y))
-  } else if (dist == "log normal") {
-    # transformation based on:
-    # https://en.wikipedia.org/wiki/Log-normal_distribution#Generation_and_parameters
-    # https://msalganik.wordpress.com/2017/01/21/making-sense-of-the-rlnorm-function-in-r/comment-page-1/
-    mean_log <- log(estimate^2 / sqrt(estimate^2 + variance))
-    sd_log <- sqrt(log(1 + (variance / estimate^2)))
-    ci <- purrr::map2(mean_log, sd_log, ~ stats::qlnorm(c(q_low, q_high), .x, .y))
-  } else if (dist == "poisson") {
-    ci <- purrr::map2(estimate, denominator, ~ get_ci_pois(q_low, q_high, .x, .y))
-  }
-
-  # wrangle output
-  result <- ci %>%
-    data.frame() %>%
-    data.table::transpose() %>%
-    dplyr::rename(lower = 1, upper = 2)
+  q_upper <- 1 - (alpha / 2)
+  q_lower <- alpha / 2
+
+  result <- data.frame(x = x, y = y) %>%
+    dplyr::mutate(
+      lower = stats::qchisq(q_lower, df = 2 * x) / (2 * y),
+      upper = stats::qchisq(q_upper, df = 2 * (x + 1)) / (2 * y)
+    ) %>%
+    dplyr::select(.data$lower, .data$upper)
 
   return(result)
 }
 
 
-#' Replicate `stats::qpois` using `stats::qchisq`
+#' Calculate a gamma distribution confidence interval
+#'
+#' @description Derive an approximate gamma distribution confidence interval
+#' using [stats::qchisq()].
+#'
+#' @param interval A scalar between 0 and 1, indicating the width of the
+#' interval. For example, use `0.95` to calculate a 95% confidence interval.
+#' @param estimate A numeric vector containing the point estimates.
+#' @param weights A list of numeric vectors containing the weights.
```

Comment:
Responding to the question in the PR, I would amend to say:

A list of numeric vectors containing the weights of each strata in the directly standardized rate calculation.

Reply:
Changed to "A list of numeric vectors where each vector contains the weights of each strata in the directly standardized rate calculation."